File: LeetCode/cpp/lc_0107.cpp
/**
 * @file     lc_0107.cpp
 * @brief    https://leetcode-cn.com/problems/binary-tree-level-order-traversal-ii/
 * @author   YongDu
 * @date     2021-09-09
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    std::vector<std::vector<int>> result;

  public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        if (!root) {
            return result;
        }
        std::queue<TreeNode*> que;
        TreeNode* node = root;
        que.push(node);
        while (!que.empty()) {
            int size = que.size();
            std::vector<int> layerNodes;
            while (size--) {
                node = que.front();
                que.pop();
                layerNodes.push_back(node->val);
                if (node->left) {
                    que.push(node->left);
                }
                if (node->right) {
                    que.push(node->right);
                }
            }
            result.emplace_back(layerNodes);
        }
        std::reverse(result.begin(), result.end());
        return result;
    }
};

File: LeetCode/javascript/lc_0107.js
/**
 * @file     levelOrderBottom.js
 * @brief    https://leetcode-cn.com/problems/binary-tree-level-order-traversal-ii/
 * @author   Zhu
 * @date     2021-07-30 12:25
 */

/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function (root) {
    let result = [];

    let queue = [];
    root && queue.push(root);
    while (queue.length > 0) {
        let levelSize = queue.length;
        let levelResult = [];
        while (levelSize-- > 0) {
            const cur = queue.shift();
            levelResult.push(cur.val);
            if (cur.left) queue.push(cur.left);
            if (cur.right) queue.push(cur.right);
        }
        result.unshift(levelResult);
    }

    return result;
};
